tests/Test/Xrefcheck/TooManyRequestsSpec.hs

```diff
@@ -0,0 +1,59 @@
+{- SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+module Test.Xrefcheck.TooManyRequestsSpec where
+
+import Universum
+
+import Control.Concurrent (forkIO, killThread)
+import qualified Control.Exception as E
+import qualified Data.CaseInsensitive as CI
+import qualified Data.Map as M
+import Network.HTTP.Types (Status (..), ok200, serviceUnavailable503, tooManyRequests429)
+import Network.HTTP.Types.Header (hRetryAfter)
+import Test.Hspec (Spec, describe, it, shouldBe)
+import Web.Firefly (ToResponse (toResponse), route, run)
+
+import Xrefcheck.Config
+import Xrefcheck.Core
+import Xrefcheck.Verify
+
+spec :: Spec
+spec = do
+  describe "429 response tests" $ do
+    it "Returns 200 eventually" $
+      checkLinkWithServer (mock429 True)
+        "http://127.0.0.1:5000/429" $ VerifyResult []
+    it "Returns 503 eventually" $
+      checkLinkWithServer (mock429 False)
+        "http://127.0.0.1:5000/429" $ VerifyResult
+          [ ExternalHttpResourceUnavailable $
+              Status { statusCode = 503, statusMessage = "Service Unavailable"}
+          ]
+  where
+    checkLinkWithServer mockServer link expectation =
+      E.bracket (forkIO (newIORef 0 >>= mockServer)) killThread $ \_ -> do
+        verifyReferenceDef link >>= (`shouldBe` expectation)
+
+    verifyReferenceDef link = do
+      progRef <- newIORef $ initVerifyProgress []
+      fmap wrlItem <$> verifyReference
+        ((cVerification $ defConfig GitHub) { vcCheckLocalhost = True }) FullMode
+        progRef (RepoInfo M.empty) link ""
+        (Reference "" (toText link) Nothing (Position Nothing))
+
+    mock429 :: Bool -> IORef Int -> IO ()
+    mock429 mode switchRef = run 5000 $ do
+      route "/429" $ do
+        switch <- atomicModifyIORef' switchRef $ \s ->
+          ((s + 1) `mod` 3, s)
+        pure $
+          if switch < 1
+          then toResponse
+            ( "" :: Text
+            , tooManyRequests429
+            , M.fromList [(CI.map (decodeUtf8 @Text) hRetryAfter, [show (switch + 1) :: Text])]
+            )
+          else toResponse ("" :: Text, if mode then ok200 else serviceUnavailable503)
```

Comment:
It boils down to one of those cases when, after having caught a bad response (like 503), the GET request is immediately performed afterward (apart from when it's our beloved 429, then the GET request is not performed). So, when `mode == False` (i.e. when the server returns a 503 response), the GET request is performed, the server then increments the `switch` to `2` and still returns 503.

Reply:
Oh! I put some traces in the server code to figure out what's happening, and I see it being called 3 times (and responds with 429 -> 503 -> 503).

I wasn't understanding why xrefcheck was making a third call - I expected xrefcheck to quit after getting a 503.

But now I remember - the 2nd call is a HEAD request!


The broader point I wanted to make was that we don't need to cycle between 0 and 2. A simple `callCount`, that is incremented on every call, would also work.

Reply:
This also reminds me: we should write a test that verifies that when a server responds to a HEAD request with 429, then xrefcheck:
* Does NOT retry a GET immediately after.
* It waits the request number of seconds and then sends another HEAD request.

Reply:
It seems to me, the only way to actually test this is to call some other verifying function, _not_ `verifyReference`, since, as far as I understand, we are not able to control & deduce what request, HEAD or GET, has been sent, because the `verifyReference`'s verification algorithm does not terminate until either we 
1. get an unfixable error or no error straight away, or
2. retry verification after catching a fixable error & get an unfixable error or no error after a certain number of retries, or
3. run out of the retries.

I believe bringing `checkHttp` to the global scope (which currently is defined in the local scope of `checkExternalResource`), and all the functions and values it depends on (like `makeHttpRequest`, `interpretErrors`, `config`, etc.) to **its** local scope, should be sufficient. By calling specifically `checkHttp` we will be able to perform 1 iteration of a link verification, granting us the ability to test how many times we've connected to the server & deduce the sent request (since, if I'm correct, there's a strict one-to-one correspondence with `callCount` and the sent requests, per single verification iteration).

Reply:
> Does NOT retry a GET immediately after

There's a simpler approach to test this: write a mock server that responds to `HEAD` requests with 429/200, but returns a 500 on `GET` requests.